streaming/compression.c

```diff
@@ -0,0 +1,343 @@
+            debug(D_STREAM, "STREAM_COMPRESSION: Destroying decompressor.");
+            if (s->data->stream)
+                LZ4_freeStreamDecode(s->data->stream);
+            freez(s->data->stream_buffer);
+        }
+        freez(s->buffer);
+        freez(s);
+        *state = NULL;
+    }
+}
+
+static size_t decode_compress_header(const char *data, size_t data_size)
+{
+    if (!data || !data_size)
+        return 0;
+    if (data_size < SIGNATURE_SIZE)
+        return 0;
+    uint32_t sign = *(uint32_t *)data;
+    if ((sign & SIGNATURE_MASK) != SIGNATURE)
+        return 0;
+    size_t length = ((sign >> 8) & 0x7f) | ((sign >> 9) & (0x7f << 7));
+    return length;
+}
+
+/*
+ * Check input data for the compression header
+ * Return the size of compressed data or 0 for uncompressed data
+ */
+size_t is_compressed_data(const char *data, size_t data_size)
+{
+    return decode_compress_header(data, data_size);
+}
+
+/*
+ * Start the collection of compressed data in an internal buffer
+ * Return the size of compressed data or 0 for uncompressed data
+ */
+static size_t lz4_decompressor_start(struct decompressor_state *state, const char *header, size_t header_size)
+{
+    size_t length = decode_compress_header(header, header_size);
+    if (!length)
+        return 0;
+
+    if (!state->buffer) {
+        state->buffer = mallocz(length);
+        state->buffer_size = length;
+    } else if (state->buffer_size < length) {
+        state->buffer = reallocz(state->buffer, length);
+        state->buffer_size = length;
+    }
+    state->buffer_len = length;
+    state->buffer_pos = 0;
+    state->out_buffer_pos = 0;
+    state->out_buffer_len = 0;
+    return length;
+}
+
+/*
+ * Add a chunk of compressed data to the internal buffer
+ * Return the current size of compressed data or 0 for error
+ */
+static size_t lz4_decompressor_put(struct decompressor_state *state, const char *data, size_t size)
+{
+    if (!state || !size || !data)
+        return 0;
+    if (!state->buffer)
+        fatal("STREAM: No decompressor buffer allocated");
+
+    if (state->buffer_pos + size > state->buffer_len) {
+        error("STREAM: Decompressor buffer overflow %lu + %lu > %lu",
+                    state->buffer_pos, size, state->buffer_len);
+        size = state->buffer_len - state->buffer_pos;
+    }
+    memcpy(state->buffer + state->buffer_pos, data, size);
+    state->buffer_pos += size;
+    return state->buffer_pos;
+}
+
+static size_t saving_percent(size_t comp_len, size_t src_len)
+{
+    if (comp_len > src_len)
+        comp_len = src_len;
+    if (!src_len)
+        return 0;
+    return 100 - comp_len * 100 / src_len;
+}
+
+/*
+ * Decompress the compressed data in the internal buffer
+ * Return the size of uncompressed data or 0 for error
+ */
+static size_t lz4_decompressor_decompress(struct decompressor_state *state)
+{
+    if (!state)
+        return 0;
+    if (!state->buffer)
+        fatal("STREAM: No decompressor buffer allocated");
```

Comment:
Is really necessary to shutdown netdata? Can we stop only the streaming channel?

Reply:
Hello @ thiagoftsm
I changed the above code according to your recommendation. Please see f7d123b.
Thank you!